### ISSUE TITLE
make minetest server multithreaded

### ISSUE THREAD

**Issue by vladtepesch** [minetest/minetest#9460]
## Problem
It seems that the minetest server is nearly single threaded.
https://dev.minetest.net/Engine_structure#Threads seems to support this assumption.

Splitting the load onto multiple threads could improve performance on slower multicore servers like raspberry pi.

Currently I get a lot of Messages that processing ABMs is to slow.
Also openeing doors sometimes has huge delays.

## Solutions

The workload of processing different parts of the world should be parallelized so that it can be processed in different threads.

**Comment by sfan5** [comment]
This isn't as easy as it sounds. For example all ABM processing requires use of the map, which needs to be synchronized (locked) and can't be used concurrently. Same goes for the Lua environment.

The Raspberry Pi is also a device that's generally a bit slow. Minetest might still not run fast enough on it even if it made better use of multiple threads.

**Comment by ryvnf** [comment]
Would it be possible to divide larger maps into different sectors (groups of chunks) which are updated independently in parallel with the rest of the world?

This would probably be a great benefit for larger servers with many active players.

I understand that this would be *a lot* of work, and is not of high priority.

**Comment by paramat** [comment]
This is such a huge and difficult task with zero chance of happening within several years and very little chance of ever happening i think the issue should be closed. The forum seems a better place.

**Comment by vladtepesch** [comment]
I am not talking about large scale server but just small one for at most 4  people. raspberry pi 4 seems to seems very desireable to be supported.



> This isn't as easy as it sounds.

adding concurrency at the end never is ^^

> map, which needs to be synchronized (locked) and can't be used concurrently.

I gues that should be the main goal at the end.
couldn't it be possible to at least independently process map blocks? Or lock based on map blocks. At least active blocks that are farther away than the active block range should be able to be processed completely independent or am I wrong? However I have no complete overview how the different things interact with each other.

Additional to this:
 a short look into the code leaves me with the impression that on many places there is unnecesary copying or use of dynamically allocated containers. However possibly these would be micro-optimizations - difficult to estimate the actual impact

**Comment by FaceDeer** [comment]
Multithreading can lead to tricky gotchas that are very hard to anticipate or debug by non-expert programmers (which many modders are, no disrespect intended).

For example, you could say that mapgen is always done on just a chunk-by-chunk basis, so why not parallelize that and let each thread work on just one chunk? But then there's a mod like "settlements", which builds towns on the map and keeps track of where each town is placed so that they aren't placed too close together. The data structure that tracks town locations isn't part of the map database, it isn't associated with any specific map chunk, so you can now get race conditions as different map chunk threads each try to check and update it at their own pace and get confused.

It might be more workable if there was some kind of specialized "register_on_mapgen_multiuthreaded" that let you register mapgen code that you are promising relies only on self-contained data structures, or an analogous "register_abm_multithreaded". I haven't really looked at Minetest's code but something like that seems to me like the lowest hanging multithreading fruit available to pick. It still depends on being able to lock just specific chunks of the map database, though, I don't know if that's easy or not. And since this would only help mods that were specifically written to take advantage of it it may not be all that big a performance gain.

**Comment by sfan5** [comment]
Lua code can never run multithreaded (at least not in the environment we have currently). Parallelizing Lua mapgens is another big task and will require modders to specifically adjust their code.

**Comment by FaceDeer** [comment]
> It still depends on being able to lock just specific chunks of the map database, though,

That about the map lock. :)

I figured since lua had support for coroutines in the base library it probably had some form of multithreading support backing it up, but if that's not in the environment that's used by Minetest then I guess you'd need to have a whole separate Lua VM running for those other threads. Never said it was _easy_, just that it seemed like the easi_est_.

**Comment by appgurueu** [comment]
Related (both closed):

* [Minetest cannot use all CPU and GPU](https://github.com/minetest/minetest/issues/9439)
* [A simple way for a better multithreading?](https://github.com/minetest/minetest/issues/4241)

**Comment by sorcerykid** [comment]
It sounds to me like the crux of the problem here is the overhead of ABMs (as usual), not overall server performance. Have you tried commenting out all registered ABMs and seeing if there's any improvement?

**Comment by paramat** [comment]
Also see #3182 #3375 #4534
Duplicate of #3375

**Comment by vladtepesch** [comment]
> 
> 
> It sounds to me like the crux of the problem here is the overhead of ABMs (as usual),
> not overall server performance. Have you tried commenting out all 
> registered ABMs and seeing if there's any improvement?

I am not completely sure about that. The server-environment loop seem to have some tasks, but only the ABM-loop has the timing-exit logic within. Maybe the other tasks simply does not appear on output.  I saw some ScopedProfiler-objects. can i activate their output with without recompiling?

Which part handles placement of objects or inventory-actions and opening doors? sometimes there is quite a large delay (>10s). mostly not. One Strange thing that i can observe is, that the cpu often is not (or just very short) at max frequency. But its definitely not an over-temperature throttle.

**Comment by sfan5** [comment]
> Which part handles placement of objects or inventory-actions and opening doors?

Interaction sent by the client is handeled in `m_server->Receive()`, eventually ends up in [`Server::handleCommand_Interact`](https://github.com/minetest/minetest/blob/master/src/network/serverpackethandler.cpp#L985) which calls `m_script->item_OnPlace` to let Lua handle the door opening.*
Inventory actions take a similar path to `Server::handleCommand_InventoryAction`.

\* The code only sends inventory changes directly. Changes to the MapBlocks are sent by `m_server->AsyncRunStep()` which is called immediately after packet processing is done.

**Comment by vladtepesch** [comment]
> Maybe the other tasks simply does not appear on output. I saw some ScopedProfiler-objects. can i activate their output with without recompiling

Please can you give me a hint?

I activated `profiler.load = true`
and used `/profiler print` and  `/profiler save` but i did not see the  `g_profiler` calls from `serverenvironment.cpp`

**Comment by sfan5** [comment]
Those settings control the Lua profiler.
To profile the engine code, press F6 (client-side) or set `profiler_print_interval` to a valid value and use the `--verbose` flag to see the output (server-side).
